File: apps/github-integration/test/contract-fixtures.ts
import { existsSync, readFileSync, readdirSync } from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'

function workspaceRoot (): string {
  let current = path.dirname(fileURLToPath(import.meta.url))
  while (true) {
    if (
      existsSync(path.join(current, 'contracts', 'README.md')) &&
      existsSync(path.join(current, 'contracts', 'fixtures'))
    ) {
      return current
    }

    const parent = path.dirname(current)
    if (parent === current) {
      throw new Error('Could not locate workspace contract fixtures root.')
    }
    current = parent
  }
}

const contractFixturesRoot = path.join(workspaceRoot(), 'contracts', 'fixtures')
const contractSchemasRoot = path.join(workspaceRoot(), 'contracts', 'schemas')

export function contractFixture (version: string, name: string): unknown {
  return JSON.parse(readFileSync(path.join(contractFixturesRoot, version, name), 'utf8'))
}

export function contractFixtureManifest (version: string): unknown {
  return JSON.parse(readFileSync(path.join(contractFixturesRoot, version, 'manifest.json'), 'utf8'))
}

export function contractFixtureFiles (version: string): string[] {
  return readdirSync(path.join(contractFixturesRoot, version))
    .filter((name) => name.endsWith('.json'))
    .sort()
}

export function contractSchema (version: string, name: string): unknown {
  return JSON.parse(readFileSync(path.join(contractSchemasRoot, version, name), 'utf8'))
}

export function contractSchemaFiles (version: string): string[] {
  return readdirSync(path.join(contractSchemasRoot, version))
    .filter((name) => name.endsWith('.schema.json'))
    .sort()
}

File: agents/tests/contract_fixtures.py
import json
from collections.abc import Iterable
from pathlib import Path
from typing import Any

from jsonschema import Draft202012Validator  # type: ignore[import-untyped]
from referencing import Registry, Resource


def _workspace_root() -> Path:
    for candidate in Path(__file__).resolve().parents:
        if (candidate / "contracts" / "README.md").is_file() and (
            candidate / "contracts" / "fixtures"
        ).is_dir():
            return candidate
    raise RuntimeError("Could not locate workspace contract fixtures root.")


CONTRACT_FIXTURES_ROOT = _workspace_root() / "contracts" / "fixtures"
CONTRACT_SCHEMAS_ROOT = _workspace_root() / "contracts" / "schemas"


def contract_fixture(version: str, name: str) -> dict[str, Any]:
    return json.loads(
        (CONTRACT_FIXTURES_ROOT / version / name).read_text(encoding="utf-8")
    )


def contract_fixture_manifest(version: str) -> dict[str, Any]:
    return json.loads(
        (CONTRACT_FIXTURES_ROOT / version / "manifest.json").read_text(encoding="utf-8")
    )


def contract_fixture_files(version: str) -> list[Path]:
    return sorted((CONTRACT_FIXTURES_ROOT / version).glob("*.json"))


def contract_schema(version: str, name: str) -> dict[str, Any]:
    return json.loads(
        (CONTRACT_SCHEMAS_ROOT / version / name).read_text(encoding="utf-8")
    )


def contract_schema_files(version: str) -> list[Path]:
    return sorted((CONTRACT_SCHEMAS_ROOT / version).glob("*.schema.json"))


def contract_schema_registry(version: str) -> Registry:
    resources: list[tuple[str, Resource]] = []
    for schema_file in contract_schema_files(version):
        schema = contract_schema(version, schema_file.name)
        resource = Resource.from_contents(schema)
        resources.append((schema["$id"], resource))
        resources.append((schema_file.name, resource))
    return Registry().with_resources(resources)


def contract_validator(
    version: str,
    schema_name: str,
) -> Draft202012Validator:
    return Draft202012Validator(
        contract_schema(version, schema_name),
        registry=contract_schema_registry(version),
    )


def format_schema_errors(errors: Iterable) -> str:
    return "\n".join(
        f"{'/'.join(str(part) for part in error.path) or '<root>'}: {error.message}"
        for error in errors
    )
